### PR TITLE
fix leveling for Kaleh and Nym in Utbs

### PR DESCRIPTION
**Pull request by newfrenchy83**
Fix https://github.com/wesnoth/wesnoth/issues/8149 issue

### CODE REVIEW

**Comment by gfgtdf** [comment]
No reason to create a new type, just use apply_to=level

**Comment by newfrenchy83** [comment]
execpt what i want use two advancement for increase level to 1, and i can done that what with change unit_type

**Comment by Hejnewar** [comment]
You can filter by xp, I know it changes with difficulty but thats fine I can always balance / fix that latter.

**Comment by newfrenchy83** [comment]
> You can filter by xp, I know it changes with difficulty but thats fine I can always balance / fix that latter.

i filtered with hp instead, sorry.

**Comment by newfrenchy83** [comment]
> You can filter by xp, I know it changes with difficulty but thats fine I can always balance / fix that latter.

it's done

**Comment by gfgtdf** [comment]
you cant to filter for `max_experience`instead of `experience`

**Comment by newfrenchy83** [comment]
> you want to filter for `max_experience`instead of `experience`



> you want to filter for `max_experience`instead of `experience`

fixed

**Comment by soliton-** [comment]
Leveling according to arbitrary XP values seems like a good way to cause future confusion when whatever this is based on is inevitably changed.

**Comment by Hejnewar** [comment]
If only there was variable or stat that contained the number of AMLAs the unit took already (maybe it is possible in lua), or if setting unit variables using advancement or effect tag was possible, this would greatly simplify a lot of things. So until that is possible I will continue using xp especially since I'm already using it for unlocking one specific Amla.

**Comment by gfgtdf** [comment]
> If only there was variable or stat that contained the number of AMLAs the unit took already (maybe it is possible in lua), or if setting unit variables using advancement or effect tag was possible

This is definitely possible and i use it a lot in my addons, but you have to write a custom effect (in lua), for example in WC2:
https://github.com/wesnoth/wesnoth/blob/master/data/campaigns/World_Conquest/lua/game_mechanics/effects.lua#L162
It's important that you use the `mods` container because that one is cleared whenever the unit is reset (so that the effects don't increase everytime the unit is reset.)

**Comment by gfgtdf** [comment]
So a possible  lua code could look like 

```
function wesnoth.effects.utbs_set_level(u, cfg)
  local n_amla = u.variables["utbs_namla"] or 0
  n_amla = n_amla + 1
  u.variables["utbs_namla"] = n_amla
  if n_amla > 2 then
    u.level = 2
  end
  if n_amla > 4 then
    u.level = 3
  end
end
```

**Comment by newfrenchy83** [comment]
If number of amla for increase level is know then value of max xp of unit just before lave-linge should could be calulated

**Comment by newfrenchy83** [comment]
> So a possible lua code could look like
> 
> ```
> function wesnoth.effects.utbs_set_level(u, cfg)
>   local n_amla = u.variables["utbs_namla"] or 0
>   n_amla = n_amla + 1
>   u.variables["utbs_namla"] = n_amla
>   if n_amla > 2 then
>     u.level = 2
>   end
>   if n_amla > 4 then
>     u.level = 3
>   end
> end
> ```
@Hejnewar 2 amla for level 2 and 2other amla for level 3, will you agreed?

**Comment by Hejnewar** [comment]
Its fine.

**Comment by newfrenchy83** [comment]
@gfgtdf could you review this PR, please?

**Comment by CelticMinstrel** [comment]
> If only there was variable or stat that contained the number of AMLAs the unit took already (maybe it is possible in lua)

Why would you need Lua for that? Isn't it just `$unit.modifications.advancement.length`? Assuming the unit is stored in `$unit` obviously.

**Comment by newfrenchy83** [comment]
> > If only there was variable or stat that contained the number of AMLAs the unit took already (maybe it is possible in lua)
> 
> Why would you need Lua for that? Isn't it just `$unit.modifications.advancement.length`? Assuming the unit is stored in `$unit` obviously.

No need to count the number of amla, just determine the maximum experience of the unit before the advancement which should make it level up, this remains simpler for the code and with better integration.

**Comment by gfgtdf** [comment]
> So a possible lua code could look like
> 
> ```
> function wesnoth.effects.utbs_set_level(u, cfg)
>   local n_amla = u.variables["utbs_namla"] or 0
>   n_amla = n_amla + 1
>   u.variables["utbs_namla"] = n_amla
>   if n_amla > 2 then
>     u.level = 2
>   end
>   if n_amla > 4 then
>     u.level = 3
>   end
> end
> ```

This code was wrong as it should use the mods container, corrected code

 ```
 function wesnoth.effects.utbs_set_level(u, cfg)
   local n_amla = u.variables["mods.utbs_namla"] or 0
   n_amla = n_amla + 1
   u.variables["mods.utbs_namla"] = n_amla
   if n_amla >= 2 then
     u.level = 2
   end
   if n_amla >= 4 then
     u.level = 3
   end
 end
 ```

**Comment by gfgtdf** [comment]
> @gfgtdf could you review this PR, please?

I'd prefer if @Hejnewar does it, as i know very little about the utbs campaign, if you have specific question about lua/wml i can answer those though.

**Comment by Hejnewar** [comment]
For me its fine if I have to review it form the campaign side, but I cant review code perspective because my lua knowledge is almost non existent. So as long as code is fine im fine as well.

**Review by Hejnewar — APPROVED** [review]

**Comment by newfrenchy83** [comment]
> For me its fine if I have to review it form the campaign side, but I cant review code perspective because my lua knowledge is almost non existent. So as long as code is fine im fine as well.

@Pentarctagon could you review this PR,please?

**Comment by Pentarctagon** [comment]
The current implementation looks like it's WFL rather than lua?

**Comment by newfrenchy83** [comment]
> The current implementation looks like it's WFL rather than lua?

yes but it is appproved, ma_experience value don't choose by chance but after level up Youth for read values for each amla

**Comment by Pentarctagon** [comment]
Well, as long as it's been tested to actually work, it looks fine to me.

**Comment by newfrenchy83** [comment]
> Well, as long as it's been tested to actually work, it looks fine to me.

it's tested in hard mode and other and this work

**Comment by Pentarctagon** [comment]
@gfgtdf @soliton- does this need to use the lua approach?

**Comment by gfgtdf** [comment]
> For me its fine if I have to review it form the campaign side, but I cant review code perspective because my lua knowledge is almost non existent. So as long as code is fine im fine as well.

The current code doesn't use any lua though.

> @gfgtdf @soliton- does this need to use the lua approach?

The current approach is not how i would do it, but I don't have a problem with merging this.

**Comment by soliton-** [comment]
Since the goal appears to be to change level after picking two AMLAs that’s what should be implemented. Manually calculating the XP that might be equivalent to this is just weird.

**Comment by CelticMinstrel** [comment]
@gfgtdf's Lua looks frankly bizarre to me. I don't understand why unit variables need to be involved in this at all.

**Comment by gfgtdf** [comment]
> @gfgtdf's Lua looks frankly bizarre to me. I don't understand why unit variables need to be involved in this at all.


Feel free to provide a better solution.

**Comment by CelticMinstrel** [comment]
Well, if taking the same basic approach of using a custom "set level" effect, I think it could look something like this:

```lua
function wesnoth.effects.utbs_set_level(u, cfg)
  local n_amla = #u.advancements + 1
  if n_amla >= 4 then
    u.level = 3
  elseif n_amla >= 2 then
    u.level = 2
  end
end
```

**Comment by gfgtdf** [comment]
> Well, if taking the same basic approach of using a custom "set level" effect, I think it could look something like this:

This approach has a problem, you modify the unit based in the state _after_ the modification (number of modifications), in practice  this causes problems for example in the following case:

1) a unit gets an amla that uses `apply_to=utbs_set_level` effect (at this point it doesn't gain a level because it only has one advancement)
2) a unit get an amla that doesnt have `apply_to=utbs_set_level` effect (maybe form some random [modification])

Now the unit is still level 1, but rebuilding (like removing an object) causes the unit to gain a level.

This example is a bit constructed because  in the campaign all advancement will have that effect, but for custom effects its still good practice to only make changes bases the previous state of the unit, to avoid running into bugs.

**Comment by CelticMinstrel** [comment]
> This example is a bit constructed because in the campaign all advancement will have that effect

Kinda feels like you're overengineering it here. If it works in the campaign, isn't that good enough?

By the way, are you implying that the unit passed to an effect already has all its modifications added to the modifications array? If that's the case it actually simplifies things a bit – there's no need to add 1 to the number of advancements.

I'd also suggest that, rather than adding this effect to every advancement, it might be easier to instead add it to an object that's placed on the unit when it's first created. There's no real need to run it once per advancement when rebuilding the unit.

**Comment by gfgtdf** [comment]
> By the way, are you implying that the unit passed to an effect already has all its modifications added to the modifications array?

I'm actually no sure on when the effect is applied for te first time, but i'm mostly talking about the point when a unit is rebuilt, because then it definitely has all modifications added to the array (even the ones not applied yes). 

> Although, that does assume that adding an advancement rebuilds the unit

It doesn't, but


imo it would be good if would give umc authors at least the change to do so, but probably out of scope for 1.18, i know that `apply_to=type` exists but it has other problems which are caused from the attempt to make [remove_object] work on these effects. In fact all suggestions except the original one one which used apply_to=type had a issue where the resiliant trait didn't work as most people would expect because the [advancement]s that set the level were applied after the trait, so at the time the trait is applied the unit is still level 1.

> Kinda feels like you're overengineering it here. If it works in the campaign, isn't that good enough?

I have, in the past dealt with various reports of "unit losing hp when removing an object" and similar issues of modifications, so i do think this is a common source of bugs.

If i have one solution that works by what i would call the default way effects currently work ("all modifications are applied in order so rebuilding the unit leads to the same result because all effects only depend on the previous unit stats") and one sulution that only works by the specific circumstances of the scenario that all amlas use that effect and that none of the other effects depend on the level, i would usually call the first solution simpler even if it involves one helper variables.

(There can of course be reasons to use the other implementation, but simplicity is not one of them)

**Comment by CelticMinstrel** [comment]
> i'm mostly talking about the point when a unit is rebuilt, because then it definitely has all modifications added to the array (even the ones not applied yes).

I think in that case it would be good to ensure that this is true also when a modification is first applied. Or, alternatively, to ensure it's _not_ true when the unit is rebuilt (and also not true when first applied, obviously). So rebuilding a unit could be implemented by essentially removing all modifications and adding them back one by one. It would have to use the `all_children_range` order for that, of course, as an object and an advancement might have depended on each other in some way in order to obtain the final stats.

**Comment by newfrenchy83** [comment]
From what I understand from this debate, is that no one seems to provide a satisfactory and simple alternative to my own method so how about you merge this PR and propose another solution later if you find one.

**Comment by soliton-** [comment]
The implementation in this PR is obscure (there is no hint what the effect is trying to achieve or where the numbers come from). If this is truly the best that can be done (which I highly doubt since Kaleh has had AMLAs since years and I'm not aware it ever been done like this before) then it seems we have a serious problem with writing robust effects that should be resolved ASAP.

**Comment by Pentarctagon** [comment]
If there is no better alternative yet, the please at least add a comment explaining the assumptions being made above https://github.com/wesnoth/wesnoth/blob/963029dd7cec3958f37e006b6a7cd1438b2d37bc/data/campaigns/Under_the_Burning_Suns/units/quenoth/Youth.cfg#L865

**Comment by newfrenchy83** [comment]
> If there is no better alternative yet, the please at least add a comment explaining the assumptions being made above
> 
> https://github.com/wesnoth/wesnoth/blob/963029dd7cec3958f37e006b6a7cd1438b2d37bc/data/campaigns/Under_the_Burning_Suns/units/quenoth/Youth.cfg#L865

comment added

File: data/campaigns/Under_the_Burning_Suns/units/quenoth/Youth.cfg [hunk 1]
```diff
@@ -862,6 +888,10 @@ _"Nym is Kaleh’s childhood friend, a young, rebellious lady with quick wits an
     movement_type=quenoth_foot
     movement=6
     hitpoints=32
+    # The evolution of this unit is carried out only with AMLAs so that it takes two amlas
+    # to reach level two then two more for level 3.
+    # As the unit variables do not include the number of amla achieved,
+    # the maximum experience is the only one that can be directly used in advancement without resorting to an external lua type function.
```
Comment (Pentarctagon) [inline data/campaigns/Under_the_Burning_Suns/units/quenoth/Youth.cfg:894]: Please also add:
`# Do not modify this value without also modifying the max_experience checks in the custom AMLAs`

Comment (newfrenchy83) [reply data/campaigns/Under_the_Burning_Suns/units/quenoth/Youth.cfg:894]: it's done

**Comment by newfrenchy83** [comment]
@Pentarctagon PR updated to integrate modification expereience because difficulty.

**Comment by Pentarctagon** [comment]
Opened #8179 for the question of how to determine the AMLA count more reliably.

**Pull request merged**